crates/ruff/src/rules/pylint/rules/cyclic_import.rs

```diff
@@ -0,0 +1,354 @@
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Cyclic import ({}) (cyclic-import)", self.cycle)
+    }
+}
+
+struct VisitedAndCycles {
+    fully_visited: FxHashSet<Arc<str>>,
+    cycles: Option<FxHashSet<Vec<Arc<str>>>>,
+}
+
+impl VisitedAndCycles {
+    fn new(fully_visited: FxHashSet<Arc<str>>, cycles: FxHashSet<Vec<Arc<str>>>) -> Self {
+        if cycles.is_empty() {
+            Self {
+                fully_visited,
+                cycles: None,
+            }
+        } else {
+            Self {
+                fully_visited,
+                cycles: Some(cycles),
+            }
+        }
+    }
+}
+
+struct CyclicImportChecker<'a> {
+    imports: &'a FxHashMap<Arc<str>, Vec<ModuleImport>>,
+}
+
+impl CyclicImportChecker<'_> {
+    fn has_cycles(&self, name: &Arc<str>) -> VisitedAndCycles {
+        let mut fully_visited: FxHashSet<Arc<str>> = FxHashSet::default();
+        let mut cycles: FxHashSet<Vec<Arc<str>>> = FxHashSet::default();
+        let mut stack: Vec<Arc<str>> = vec![name.clone()];
+        self.has_cycles_helper(name, &mut stack, &mut cycles, &mut fully_visited, 0);
+        VisitedAndCycles::new(fully_visited, cycles)
+    }
+
+    fn has_cycles_helper(
+        &self,
+        name: &Arc<str>,
+        stack: &mut Vec<Arc<str>>,
+        cycles: &mut FxHashSet<Vec<Arc<str>>>,
+        fully_visited: &mut FxHashSet<Arc<str>>,
+        level: usize,
+    ) {
+        if let Some(imports) = self.imports.get(name) {
+            let tabs = "\t".repeat(level);
+            debug!("{tabs}{name}");
+            for import in imports.iter() {
+                debug!("{tabs}\timport: {}", import.module);
+                if let Some(idx) = stack.iter().position(|s| s == &import.module) {
+                    debug!("{tabs}\t\t cycles: {:?}", stack[idx..].to_vec());
+                    // when the length is 1 and the only item is the import we're looking at
+                    // then we're importing self, could we report this so we don't have to
+                    // do this again for import-self W0406?
+                    if stack[idx..].len() == 1 && stack[idx] == import.module {
+                        continue;
+                    }
+                    cycles.insert(stack[idx..].to_vec());
+                } else {
+                    stack.push(import.module.clone());
+                    self.has_cycles_helper(&import.module, stack, cycles, fully_visited, level + 1);
+                    stack.pop();
+                }
+            }
+        }
+        fully_visited.insert(name.clone());
+    }
+}
+
+/// PLR0914
+pub fn cyclic_import(
+    path: &Path,
+    package: Option<&Path>,
+    imports: &ImportMap,
+    cycles: &mut FxHashMap<Arc<str>, FxHashSet<Vec<Arc<str>>>>,
+) -> Option<Vec<Diagnostic>> {
+    let Some(package) = package else {
+        return None;
+    };
+    let module_name = to_module_path(package, path).unwrap().join(".");
```

Comment:
Can we document why calling `unwrap` is safe or avoid calling `unwrap`?

Reply:
Sure, I'll need to double check, but I think that at this point, because this function was called to do the exact same thing one the same arguments when creating the `ImportMap`, then it must be safe here. I will need to check that is unequivocally true.

Reply:
I don't _believe_ it's unsafe to call `unwrap` here, however to be certain, I've avoided it now.